src/app/shared/context.service.ts

```diff
@@ -92,17 +94,59 @@ export class ContextService implements Contexts {
       })
       .multicast(() => new ReplaySubject(1));
     // Subscribe the the default context to the recent space collector
-    this._default.subscribe(addRecent);
+    this._default.subscribe(this._addRecent);
 
-    // Compute the current context
-    this._current = this.broadcaster.on<Navigation>('navigate')
-      // Eliminate duplicate navigation events
-      // TODO this doesn't work quite perfectly
-      .distinctUntilKeyChanged('url')
-      // Extract the user and space names from the URL
-      .map(val => {
-        return { user: this.extractUser(), space: this.extractSpace(), url: val.url } as RawContext;
+    // Create the recent space list
+    this._recent = this._addRecent
+      // Map from the context being added to an array of recent contexts
+      // The scan operator allows us to access the list of recent contexts and add ours
+      .scan((recent, ctx) => {
+        // First, check if this context is already in the list
+        // If it is, remove it, so we don't get duplicates
+        for (let i = recent.length - 1; i >= 0; i--) {
+          if (recent[i].path === ctx.path) {
+            recent.splice(i, 1);
+          }
+        }
+        // Then add this context to the top of the list
+        recent.unshift(ctx);
+        return recent;
+        // The final value to scan is the initial value, used when the app starts
+      }, this.dummy.recent)
+      .do(val => {
+        // Truncate the number of recent contexts to the correct length
+        if (val.length > this.RECENT_CONTEXT_LENGTH) {
+          val.splice(
+            this.RECENT_CONTEXT_LENGTH,
+            val.length - this.RECENT_CONTEXT_LENGTH
+          );
+        }
+      })
+      // Finally save the list of recent contexts
+      .do(val => {
+        this.dummy.recent = val;
+        this.broadcaster.broadcast('save');
       })
+      .multicast(() => new ReplaySubject(1));
+    // Finally, start broadcasting
+    this._default.connect();
+    this._recent.connect();
+  }
+
+  get recent(): Observable<Context[]> {
+    return this._recent;
+  }
+
+  get current(): Observable<Context> {
+    return this._current;
+  }
+
+  get default(): Observable<Context> {
```

Comment:
This isn't a change, it's just a moved method. See https://github.com/fabric8-ui/ngx-fabric8-wit/issues/34